### PR TITLE
Test to ensure that remote CCS and remote CCR actions don't overlap

### PR DESCRIPTION
**Pull request by jakelandis**
This commit introduces a test to help ensure that actions used between CCR and CCS are evaluated on case-by-case basis if/when the allowed permissions overlap. Some cases are fine if they overlap, other cases it could be a security bug. This will test will fail the permissions we assign to the RCS 2.0 API key overlap.

### CODE REVIEW

**Comment by jakelandis** [comment]
pulled back to draft ... the test introduced here failed in CI but not locally (and no randomization). need to better understand why

**Comment by jakelandis** [comment]
@elastic/es-delivery - could you take a look at the introduction of a new test task (in xpack core). The reason I need this is that for testing only reasons, I need to record additional context to what the production code is doing, and the production code is in static state. So the new test task uses a test only JVM argument...trying to keep the change here isolated to only changing production code behavior for this 1 test.

File: x-pack/plugin/core/build.gradle [hunk 1]
```diff
@@ -176,3 +176,16 @@ if (BuildParams.inFipsJvm){
   // Test clusters run with security disabled
   tasks.named("javaRestTest").configure{enabled = false }
 }
+
+//this specific test requires a test only system property to be set, so we run it in a different JVM via a separate task
+tasks.register('testAutomatonPatterns', Test) {
+  include '**/AutomatonPatternsTests.class'
+  systemProperty 'tests.automaton.record.patterns', 'true'
```
Comment (mark-vieira) [inline x-pack/plugin/core/build.gradle:183]: I assume it's not ok to just set this system property for all other tests as well?

**Review by mark-vieira — APPROVED** [review]

File: x-pack/plugin/core/build.gradle (hunk 1 above)
Comment (jakelandis) [reply x-pack/plugin/core/build.gradle:183]: It probably wouldn't hurt to run all of our tests with this set for all of them, but then we would be running all of our tests in a configuration that does not match how production code is ever run. We could also just change production code to always record the patterns, but it adds a tiny bit of overhead that is only ever used by this one test. The special test task here is a bit awkward, but IMO is better than alternatives.

Comment (mark-vieira) [reply x-pack/plugin/core/build.gradle:183]: Got it, I wasn't sure if it was completely benign and nonapplicable to the other tests runtime behavior.

**Review by n1v0lg — COMMENTED** [review]
I'm good with this approach, but wanted to suggest an alternative that avoids the system property flag and "runtime" behavior toggle. It also gets rid of the need for a specialized task in gradle:

```diff
diff --git a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java
index 8443dc40836..62929e17f9d 100644
--- a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java
+++ b/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java
@@ -76,12 +76,14 @@ public final class IndexPrivilege extends Privilege {
     private static final Logger logger = LogManager.getLogger(IndexPrivilege.class);
 
     private static final Automaton ALL_AUTOMATON = patterns("indices:*", "internal:transport/proxy/indices:*");
-    private static final Automaton READ_AUTOMATON = patterns(
+    // pck-private for testing
+    static final AutomatonWithPatterns READ_AUTOMATON = new AutomatonWithPatterns(
         "indices:data/read/*",
         ResolveIndexAction.NAME,
         TransportResolveClusterAction.NAME
     );
-    private static final Automaton READ_CROSS_CLUSTER_AUTOMATON = patterns(
+    // pck-private for testing
+    static final AutomatonWithPatterns READ_CROSS_CLUSTER_AUTOMATON = new AutomatonWithPatterns(
         "internal:transport/proxy/indices:data/read/*",
         TransportClusterSearchShardsAction.TYPE.name(),
         TransportSearchShardsAction.TYPE.name(),
@@ -89,6 +91,13 @@ public final class IndexPrivilege extends Privilege {
         "indices:data/read/esql",
         "indices:data/read/esql/compute"
     );
+
+    record AutomatonWithPatterns(Automaton automaton, String... patterns) {
+        AutomatonWithPatterns(String... patterns) {
+            this(Automatons.patterns(patterns), patterns);
+        }
+    }
+
     private static final Automaton CREATE_AUTOMATON = patterns(
         "indices:data/write/index*",
         "indices:data/write/bulk*",
@@ -122,7 +131,8 @@ public final class IndexPrivilege extends Privilege {
         CreateDataStreamAction.NAME
     );
     private static final Automaton DELETE_INDEX_AUTOMATON = patterns(TransportDeleteIndexAction.TYPE.name(), DeleteDataStreamAction.NAME);
-    private static final Automaton VIEW_METADATA_AUTOMATON = patterns(
+    // pck-private for testing
+    static final AutomatonWithPatterns VIEW_METADATA_AUTOMATON = new AutomatonWithPatterns(
         GetAliasesAction.NAME,
         GetIndexAction.NAME,
         GetFieldMappingsAction.NAME + "*",
@@ -178,8 +188,11 @@ public final class IndexPrivilege extends Privilege {
 
     public static final IndexPrivilege NONE = new IndexPrivilege("none", Automatons.EMPTY);
     public static final IndexPrivilege ALL = new IndexPrivilege("all", ALL_AUTOMATON);
-    public static final IndexPrivilege READ = new IndexPrivilege("read", READ_AUTOMATON);
-    public static final IndexPrivilege READ_CROSS_CLUSTER = new IndexPrivilege("read_cross_cluster", READ_CROSS_CLUSTER_AUTOMATON);
+    public static final IndexPrivilege READ = new IndexPrivilege("read", READ_AUTOMATON.automaton());
+    public static final IndexPrivilege READ_CROSS_CLUSTER = new IndexPrivilege(
+        "read_cross_cluster",
+        READ_CROSS_CLUSTER_AUTOMATON.automaton()
+    );
     public static final IndexPrivilege CREATE = new IndexPrivilege("create", CREATE_AUTOMATON);
     public static final IndexPrivilege INDEX = new IndexPrivilege("index", INDEX_AUTOMATON);
     public static final IndexPrivilege DELETE = new IndexPrivilege("delete", DELETE_AUTOMATON);
@@ -189,7 +202,7 @@ public final class IndexPrivilege extends Privilege {
     public static final IndexPrivilege MANAGE = new IndexPrivilege("manage", MANAGE_AUTOMATON);
     public static final IndexPrivilege DELETE_INDEX = new IndexPrivilege("delete_index", DELETE_INDEX_AUTOMATON);
     public static final IndexPrivilege CREATE_INDEX = new IndexPrivilege("create_index", CREATE_INDEX_AUTOMATON);
-    public static final IndexPrivilege VIEW_METADATA = new IndexPrivilege("view_index_metadata", VIEW_METADATA_AUTOMATON);
+    public static final IndexPrivilege VIEW_METADATA = new IndexPrivilege("view_index_metadata", VIEW_METADATA_AUTOMATON.automaton());
     public static final IndexPrivilege MANAGE_FOLLOW_INDEX = new IndexPrivilege("manage_follow_index", MANAGE_FOLLOW_INDEX_AUTOMATON);
     public static final IndexPrivilege MANAGE_LEADER_INDEX = new IndexPrivilege("manage_leader_index", MANAGE_LEADER_INDEX_AUTOMATON);
     public static final IndexPrivilege MANAGE_ILM = new IndexPrivilege("manage_ilm", MANAGE_ILM_AUTOMATON);
diff --git a/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilegeTests.java b/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilegeTests.java
index 265714ee6ea..e85ecd0f8f2 100644
--- a/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilegeTests.java
+++ b/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilegeTests.java
@@ -17,17 +17,26 @@ import org.elasticsearch.action.update.TransportUpdateAction;
 import org.elasticsearch.common.util.iterable.Iterables;
 import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.xpack.core.rollup.action.GetRollupIndexCapsAction;
+import org.elasticsearch.xpack.core.security.action.apikey.CrossClusterApiKeyRoleDescriptorBuilder;
 import org.elasticsearch.xpack.core.transform.action.GetCheckpointAction;
 
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.List;
+import java.util.Locale;
 import java.util.Set;
 
+import static java.util.stream.Collectors.toList;
+import static org.elasticsearch.xpack.core.security.authz.privilege.IndexPrivilege.READ_AUTOMATON;
+import static org.elasticsearch.xpack.core.security.authz.privilege.IndexPrivilege.READ_CROSS_CLUSTER_AUTOMATON;
+import static org.elasticsearch.xpack.core.security.authz.privilege.IndexPrivilege.VIEW_METADATA_AUTOMATON;
 import static org.elasticsearch.xpack.core.security.authz.privilege.IndexPrivilege.findPrivilegesThatGrant;
 import static org.hamcrest.Matchers.containsInAnyOrder;
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasItem;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.lessThan;
+import static org.hamcrest.Matchers.not;
 
 public class IndexPrivilegeTests extends ESTestCase {
 
@@ -57,6 +66,45 @@ public class IndexPrivilegeTests extends ESTestCase {
         assertThat(viewIndexMetadata, lessThan(manage));
     }
 
+    public void testCcsAndCcrPrivilegesOverlap() {
+        // assert we are using a complete set of CCS automata that cover all CCS index privileges
+        var ccsAutomataWithPatterns = List.of(READ_AUTOMATON, READ_CROSS_CLUSTER_AUTOMATON, VIEW_METADATA_AUTOMATON);
+        assertThat(
+            ccsAutomataWithPatterns.stream().map(IndexPrivilege.AutomatonWithPatterns::automaton).collect(toList()),
+            containsInAnyOrder(
+                Arrays.stream(CrossClusterApiKeyRoleDescriptorBuilder.CCS_INDICES_PRIVILEGE_NAMES)
+                    .map(name -> IndexPrivilege.getNamedOrNull(name).getAutomaton())
+                    .toArray()
+            )
+        );
+
+        var ccrPrivileges = CrossClusterApiKeyRoleDescriptorBuilder.CCR_INDICES_PRIVILEGE_NAMES;
+        // TODO assert we are using a complete set of CCR automata that cover all CCR index privileges
+
+        // assert no overlap for CCS
+        for (var ccsAutomatonWithPatterns : ccsAutomataWithPatterns) {
+            for (var ccsPatterns : ccsAutomatonWithPatterns.patterns()) {
+                // TODO potentially still need "ccsPattern.replaceAll("\\*", randomAlphaOfLengthBetween(1, 8));" here
+                var grantingPrivileges = Set.copyOf(findPrivilegesThatGrant(ccsPatterns));
+                for (var ccrPrivilege : ccrPrivileges) {
+                    assertThat(
+                        String.format(
+                            Locale.ROOT,
+                            "CCR privilege \"%s\" allows CCS action \"%s\". This could result in an "
+                                + "accidental bleeding of permission between RCS 2.0's search and replication index permissions",
+                            ccrPrivilege,
+                            ccsPatterns
+                        ),
+                        grantingPrivileges,
+                        not(hasItem(ccrPrivilege))
+                    );
+                }
+            }
+        }
+
+        // TODO assert no overlap for CCR
+    }

```

The idea is to "wrap" the automata we are interested in with a record that captures both automaton and raw patterns and use that in a unit test.

IMO this is a bit simpler and less "invasive" though I don't feel strongly and am happy to also go with the original approach.

**Comment by jakelandis** [comment]
>  but wanted to suggest an alternative that avoids the system property flag and "runtime" behavior toggle. It also gets rid of the need for a specialized task in gradle

Thanks for the suggestion, but I prefer the existing approach. I don't want to create 2 types of automatons or require the (extremely minimal) additional memory footprint across all running ES nodes just for this test. I also think having easy access to this for all automatons might help with future tests without the need to change any production code. The setup is kinda awkward, but should also be transparent and isolated to development only work.

**Review by n1v0lg — APPROVED** [review]
LGTM 👍

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/Automatons.java [hunk 2]
```diff
@@ -87,10 +95,13 @@ public static Automaton patterns(Collection<String> patterns) {
             return EMPTY;
         }
         if (cache == null) {
-            return buildAutomaton(patterns);
+            return maybeRecordPatterns(buildAutomaton(patterns), patterns);
```
[the remark this answers is not in the document]
Comment (jakelandis) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/Automatons.java:98]: not sure I follow... everywhere (the 2 places) that `buildAutomaton` is called, it is wrapped in `maybeRecordPatterns`, and the `patterns(String... patterns)` just delegates to `patterns(Collection<String> patterns)`.

[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/Automatons.java:98]: Ah I meant [here](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/Automatons.java#L195-L205)

[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/Automatons.java:98]: Looking again though, we'd just end up calling `maybeRecordPatterns` multiple times. All good to leave as is.

**Pull request merged**